File: frontend/src/types/temporal.ts
export type HealthTrendClassification = 'STABLE' | 'IMPROVING' | 'DECLINING' | 'RAPIDLY_DECLINING';

export interface ZoneScanRecord {
  scan_id: string;
  observation_date: string;
  health_score: number; // 0-100
  mean_ndvi?: number;
  mean_ndre?: number;
  disease_probability: number; // 0-100
  disease_count: number;
  max_disease_severity: string;
  pest_probability: number; // 0-100
  pest_count: number;
  water_stress_cwsi: number; // 0.0 - 1.0
  water_stress_category: string;
  anomaly_score: number; // 0-100
}

export interface ZoneHistoryResponse {
  zone_id: string;
  zone_code: string;
  farm_id: string;
  total_scans: number;
  history: ZoneScanRecord[];
}

export interface ZoneTrendResponse {
  zone_id: string;
  zone_code: string;
  farm_id: string;
  total_scans: number;
  health_trend: HealthTrendClassification;
  disease_trend: string;
  pest_trend: string;
  water_stress_trend: string;
  anomaly_trend: string;
  classification: HealthTrendClassification;
  delta_last_scan_percent: number;
  delta_total_percent: number;
  velocity_per_day: number;
  baseline_health: number;
  current_health: number;
  recent_scans: ZoneScanRecord[];
}

export interface FarmHealthTimelineSnapshot {
  date: string;
  mean_health_score: number;
  mean_ndvi?: number;
  total_observations: number;
  zones_monitored: number;
  dominant_trend: string;
}

export interface FarmHealthTimelineResponse {
  farm_id: string;
  farm_name: string;
  total_zones: number;
  history_by_date: FarmHealthTimelineSnapshot[];
}

File: backend/app/schemas/temporal.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field

from app.schemas.common import BaseSchema


class ZoneScanRecord(BaseSchema):
    """Normalized metric record from a single drone monitoring flight scan."""

    scan_id: str
    observation_date: datetime
    health_score: float = Field(description="Canopy health score (0-100)")
    mean_ndvi: Optional[float] = None
    mean_ndre: Optional[float] = None
    disease_probability: float = Field(default=0.0, description="Disease pathology probability (0-100)")
    disease_count: int = 0
    max_disease_severity: str = "none"
    pest_probability: float = Field(default=0.0, description="Pest risk probability (0-100)")
    pest_count: int = 0
    water_stress_cwsi: float = Field(default=0.0, description="Crop Water Stress Index (0.0 - 1.0)")
    water_stress_category: str = "none"
    anomaly_score: float = Field(default=0.0, description="Canopy spatial anomaly percentage (0-100)")


class ZoneHistoryResponse(BaseSchema):
    """Chronological observation history for a specific monitoring zone."""

    zone_id: str
    zone_code: str
    farm_id: str
    total_scans: int
    history: List[ZoneScanRecord]


class ZoneTrendResponse(BaseSchema):
    """Calculated temporal trend metrics, rate of change, and classification for a zone."""

    zone_id: str
    zone_code: str
    farm_id: str
    total_scans: int
    health_trend: str = Field(description="STABLE, IMPROVING, DECLINING, RAPIDLY_DECLINING")
    disease_trend: str = Field(description="STABLE, IMPROVING, INCREASING, SURGING")
    pest_trend: str = Field(description="STABLE, IMPROVING, INCREASING, SURGING")
    water_stress_trend: str = Field(description="STABLE, RELIEVED, INCREASING, CRITICAL")
    anomaly_trend: str = Field(description="STABLE, DISSIPATING, EXPANDING")
    classification: str = Field(description="STABLE, IMPROVING, DECLINING, RAPIDLY_DECLINING")
    delta_last_scan_percent: float
    delta_total_percent: float
    velocity_per_day: float
    baseline_health: float
    current_health: float
    recent_scans: List[ZoneScanRecord]


class FarmHealthTimelineSnapshot(BaseSchema):
    """Farm-level aggregate health trajectory snapshot on a specific date."""

    date: str
    mean_health_score: float
    mean_ndvi: Optional[float] = None
    total_observations: int
    zones_monitored: int
    dominant_trend: str


class FarmHealthTimelineResponse(BaseSchema):
    """Farm-level temporal health trajectory across all subdivided zones."""

    farm_id: str
    farm_name: str
    total_zones: int
    history_by_date: List[FarmHealthTimelineSnapshot]
